tools/python/hdf5_lib.py

```diff
@@ -0,0 +1,1106 @@
+        pilot_sc_ind = np.array(metadata['OFDM_PILOT_SC'])
+        pilot_sc_val = np.array(metadata['OFDM_PILOT_SC_VALS'])
+        zero_sc_ind = np.setdiff1d(range(fft_size), data_sc_ind)
+        zero_sc_ind = np.setdiff1d(zero_sc_ind, pilot_sc_ind)
+        nonzero_sc_ind = np.setdiff1d(range(fft_size), zero_sc_ind)
+        modulation = metadata['CL_MODULATION'].astype(str)
+        ofdm_obj = ofdmTxRx()
+        data_sc_len = len(data_sc_ind)
+
+
+        # UL Samps: #Frames, #Uplink SLOTS, #Antennas, #Samples
+        n_frames = ul_samps.shape[0]
+        n_slots = ul_samps.shape[1]
+        n_ants = ul_samps.shape[2]
+        n_users = csi.shape[1]
+        ul_syms = np.empty((n_frames, n_slots, n_ants,
+                       symbol_per_slot, fft_size), dtype='complex64')
+
+
+        # UL Syms: #Frames, #Uplink SLOTS, #Antennas, #OFDM Symbols, #Samples
+        for i in range(symbol_per_slot):
+            ul_syms[:, :, :, i, :] = ul_samps[:, :, :, offset + cp + i*ofdm_len:offset+(i+1)*ofdm_len]
+        if fft_shifted_dataset:
+            ul_syms_f = np.fft.fftshift(np.fft.fft(ul_syms, fft_size, 4), 4)
+        else:
+            ul_syms_f = np.fft.fft(ul_syms, fft_size, 4)
+
+        ul_equal_syms = np.zeros((n_frames, n_users, symbol_per_slot * data_sc_len), dtype='complex64')
+
+        # process tx data
+        rep = n_frames // txdata.shape[0]
+        tx_symbols = np.tile(txdata, (rep, 1, 1, 1, 1))
+        frac_fr = n_frames % txdata.shape[0]
+        #tx_symbols = np.tile(txdata[:1], (n_frames, 1, 1, 1))
+        #frac_fr = 0
+        if frac_fr > 0:
+            frac = txdata[:frac_fr, :, :, :]
+            tx_symbols = frac if rep == 0 else np.concatenate((tx_symbols, frac), axis=0)
+        tx_data_syms = np.reshape(tx_symbols[:, :, :, :, data_sc_ind], (tx_symbols.shape[0], tx_symbols.shape[1], tx_symbols.shape[2], symbol_per_slot * data_sc_len))
+        useful_frame_num = tx_data_syms.shape[0]
+        if txdata.shape[0] > 1:
+            useful_frame_num = useful_frame_num - max(ue_frame_offset)
+        min_ue_offset = min(ue_frame_offset)
+        slot_evm = np.zeros((useful_frame_num, n_users))
+        slot_evm_snr = np.zeros((useful_frame_num, n_users))
+        slot_ser = np.zeros((useful_frame_num, n_users))
+
+        M = 2
+        if modulation == 'QPSK':
+            M = 4
+        elif modulation == '16QAM':
+            M = 16
+        elif modulation == '64QAM':
+            M = 64
+
+        if method != 'ml':
+            ul_syms_f_tp = np.transpose(ul_syms_f[:, :, :, :, nonzero_sc_ind], (0, 1, 3, 2, 4))
+            # UL DEMULT: #Frames, #OFDM Symbols, #User, #Sample (DATA + PILOT SCs)
+            # We assume the first two slots are pilots used for phase tracking
+            # Slot 2 onwards in Agora are data 
+            ul_demult0 = demult(csi, ul_syms_f_tp[:, 0, :, :, :], noise_samps_f, method=method)
+            phase_corr = np.sum(ul_demult0 * np.conj(ue_pilot), axis=3)
```

Comment:
This phase correction is for Agora only (not Sounder). Should we keep it separate from Sounder or have a script that supports both?